### PR TITLE
ESPnet-Spk part 3 - inference every epoch using EER

### PR DESCRIPTION
**Pull request by Jungjee**
This PR is a follow-up PR on #5258.
We only had training available, whereas through this PR, we finish the ESPnet-SPK basic pipeline.
Now, EER is calculated every epoch.

### CODE REVIEW

File: egs2/mini_an4/spk1/local/make_trial.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    with open(sys.argv[1]) as f:
+        lines = f.readlines()
+
+    joint_key = lines[0].strip().split(" ")[0] + "*" + lines[1].strip().split(" ")[0]
+    with open(os.path.join(sys.argv[2], "trial.scp"), "w") as f:
+        f.write(joint_key+ " " + " ".join(lines[0].strip().split(" ")[1:]))
+    with open(os.path.join(sys.argv[2], "trial2.scp"), "w") as f:
+        f.write(joint_key+ " " + " ".join(lines[1].strip().split(" ")[1:]))
+    with open(os.path.join(sys.argv[2], "trial_label"), "w") as f:
+        f.write(joint_key + " 0")
```
Comment (Emrys365) [inline egs2/mini_an4/spk1/local/make_trial.py]: You need to add linebreaks "\n" in each `f.write`.

File: egs2/mini_an4/spk1/local/make_trial.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import sys, os
+
+if __name__ == "__main__":
+    with open(sys.argv[1]) as f:
+        lines = f.readlines()
+
+    joint_key = lines[0].strip().split(" ")[0] + "*" + lines[1].strip().split(" ")[0]
+    with open(os.path.join(sys.argv[2], "trial.scp"), "w") as f:
```
Comment (Emrys365) [inline egs2/mini_an4/spk1/local/make_trial.py:9]: Maybe it's better to rename `trial.scp` to `trial1.scp` to make them more consistent.

File: egs2/mini_an4/spk1/local/make_trial.py (hunk 1 above)
Comment (Jungjee) [reply egs2/mini_an4/spk1/local/make_trial.py]: yeah, we discussed this and I already reflected ! thanks!

**Comment by Jungjee** [comment]
Great! current version passes all CI.
@Emrys365, @sw005320, let's aim for merging this pr as well within today. 

I will also add another one PR (will be small compared to this) to add some metric learning objective function (thinking of SimCLR style contrastive at the moment).

**Comment by sw005320** [comment]
Can you fix the conflicts due to the merge of your previous PRs?

**Comment by Jungjee** [comment]
Just resolved all conflicts! Now same to my local branch which passes CI tests

File: egs2/TEMPLATE/spk1/spk.sh [hunk 3]
```diff
@@ -181,21 +181,75 @@ if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
...
+        fi
+
+        # Calculate EER for valid/test since speaker verification is an open set problem
+        # Train can be either multi-column data or not, but valid/test always require multi-column trial
         for dset in ${_dsets}; do
             utils/copy_data_dir.sh --validate_opts --non-print data/"${dset}" "${data_feats}/${dset}"
-            echo "${feats_type}" > "${data_feats}/${dset}/feats_type"
+            cp data/${dset}/trial.scp "${data_feats}/${dset}"
+            cp data/${dset}/trial2.scp "${data_feats}/${dset}"
+            cp data/${dset}/trial_label "${data_feats}/${dset}"
```
Comment (ftshijt) [inline egs2/TEMPLATE/spk1/spk.sh:212]: Are these lines necessary given that you will generate trial.scp in following format_wav_scp?

File: egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  sample_rate: 16000
+  num_eval: 5
+
+model_conf:
+  extract_feats_in_collect_stats: false
+
+loss: aamsoftmax
+loss_conf:
+  nOut: 256
+  nClasses: 7205
```
Comment (ftshijt) [inline egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr.yaml:34]: It is OK to keep arguments like the current format to keep it consistent with the source codebase, but is it possible to make consistent to other ESPnet style module? (making it to output_size and class_num

**Review by ftshijt — COMMENTED** [review]
Great effort! some minor comments.

File: egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr.yaml (hunk 4 above)
Comment (sw005320) [reply egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr.yaml:34]: I agree.
It would be better to unify it.
At least, it should be consistent with the config variable naming convention (lower case, "_" for the space)

File: espnet2/iterators/sequence_iter_factory.py [hunk 5]
```diff
@@ -140,6 +143,18 @@ def build_iter(self, epoch: int, shuffle: bool = None) -> DataLoader:
         else:
             kwargs = {}
 
+        # reshuffle whole 'batches' so that elements within a batch can move
+        # between different batches
+        if self.shuffle_within_batch:
```
Comment (sw005320) [inline espnet2/iterators/sequence_iter_factory.py:148]: Will it change the result?
If we are using the basic loss, the pooling operation across the sentence is order-agnostic, and it would not change the result (except for the numerical difference).

Comment (sw005320) [reply espnet2/iterators/sequence_iter_factory.py:148]: Let's try to pass this part by a test.

File: espnet2/schedulers/cosine_anneal_warmup_restart.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
+# code from: https://github.com/katsura-jp/pytorch-cosine-annealing-with-
```
Comment (sw005320) [inline espnet2/schedulers/cosine_anneal_warmup_restart.py:1]: add a test

File: espnet2/tasks/spk.py [hunk 7]
```diff
@@ -252,7 +252,8 @@ def required_data_names(
     def optional_data_names(
         cls, train: bool = True, inference: bool = False
     ) -> Tuple[str, ...]:
-        retval = ("spk_labels",)
+        retval = ("speech2", "trial", "spk_labels")
```
Comment (sw005320) [inline espnet2/tasks/spk.py:255]: what is "speech2"?

File: espnet2/train/preprocessor.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 import re
```
Comment (sw005320) [inline espnet2/train/preprocessor.py:4]: add a test to incorporate the data augmentation

File: espnet2/train/spk_trainer.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
+"""Trainer module for speaker recognition."""
```
Comment (sw005320) [inline espnet2/train/spk_trainer.py:1]: ditto
add a test

File: espnet2/train/spk_trainer.py
Comment (sw005320) [inline espnet2/train/spk_trainer.py:1]: what is `speech2`?

Comment (sw005320) [reply espnet2/train/spk_trainer.py:1]: - is there a better name?
- will we have `speech3`?

File: espnet2/utils/eer.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
+"""
```
Comment (sw005320) [inline espnet2/utils/eer.py:1]: ditto
add a test.

It is especially important since the objective scoring MUST NOT have any errors.

**Review by sw005320 — COMMENTED** [review]
I added several comments.
I think the overall structure looks fine.
Please keep on working on it!
We're almost there.

File: egs2/TEMPLATE/spk1/spk.sh (hunk 3 above)
Comment (Jungjee) [reply egs2/TEMPLATE/spk1/spk.sh:212]: makes sense! Thanks for pointing out. I removed the first two, which does go through format_wav_scp

File: egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr.yaml (hunk 4 above)
Comment (Jungjee) [reply egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr.yaml:34]: done!

File: egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr_bs.yaml [hunk 11]
```diff
@@ -0,0 +1,60 @@
+# RawNet3 reproduce recipe configuration.
```
[the remark this answers is not in the document]
Comment (Jungjee) [reply egs2/voxceleb/spk1/conf/tuning/train_RawNet3_sgdr_bs.yaml:1]: it was just tuning. I removed some unnecessary configs and made current best as a symlink out of `tuning` folder

File: espnet2/iterators/sequence_iter_factory.py (hunk 5 above)
Comment (Jungjee) [reply espnet2/iterators/sequence_iter_factory.py:148]: For speaker recognition, having diverse speaker classes within each mini-batch is critical. 
I looked at many ESPnet dataset code, but found that `batch_sampler` in ESPnet only plays with the order of minibatchs between epochs but don't randomly shuffle the samples that goes into each mini-batch.

I believe this is related to the nature of ASR and some other tasks where we sort in length. 
In contrast, in speaker recognition, we need same length utterances within the mini-batch and also needs to belong to different speakers. 

I empirically found that without `shuffle_within_batch`, even the training loss doesn't decrease at all.

File: espnet2/tasks/spk.py (hunk 7 above)
Comment (Jungjee) [reply espnet2/tasks/spk.py:262]: When calculating EER, we need trials where each trial has two utterances. 
`speech2` is for the second utterance of a trial pair.

Any recommendations for an alternative name?

File: espnet2/train/spk_trainer.py
Comment (Jungjee) [reply espnet2/train/spk_trainer.py:1]: based on my explanation above, there will not be speech3 or 4

File: espnet2/tasks/spk.py (hunk 7 above)
Comment (sw005320) [reply espnet2/tasks/spk.py:262]: With this explanation in the comment, you can keep using this.

Comment (Jungjee) [reply espnet2/tasks/spk.py:262]: ok, then I will add comment on this

File: espnet2/train/spk_trainer.py (hunk 9 above)
Comment (Jungjee) [reply espnet2/train/spk_trainer.py:1]: This one' already included in ci/test_integration2.sh! :)

**Comment by kan-bayashi** [comment]
Hi @Jungjee, Could you minimize the parameters for mini_an4 as possible as you can ?
You can refer https://github.com/espnet/espnet/pull/5324

**Comment by Jungjee** [comment]
@kan-bayashi , the number of parameter of ESPnet-SPK's ci model was 0.1M and I further reduced it to 0.01M. Hope this is small enough. Let me know if you need further alterations.

File: espnet2/spk/loss/aamsoftmax.py [hunk 12]
```diff
@@ -21,24 +21,24 @@ class AAMSoftmax(AbsLoss):
     vision and pattern recognition. 2019.
 
     args:
-        nOut    : dimensionality of speaker embedding
-        nClaases: number of speakers in the training set
+        nout    : dimensionality of speaker embedding
+        nclaases: number of speakers in the training set
```
Comment (Emrys365) [inline espnet2/spk/loss/aamsoftmax.py:25]:
```suggestion
        nclases: number of speakers in the training set
```

File: espnet2/train/spk_trainer.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    @contextmanager
+    def autocast(enabled=True):
+        yield
+
+    GradScaler = None
+
+try:
+    import fairscale
+except ImportError:
+    fairscale = None
```
Comment (Emrys365) [inline espnet2/train/spk_trainer.py:64]:
```suggestion
```

Comment (Emrys365) [reply espnet2/train/spk_trainer.py]: You can remove these lines if they are not used.

File: espnet2/train/spk_trainer.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+        # exception for collect_stats.
+        if len(scores) == 1:
+            reporter.register(stats=dict(eer=1.0, mindcf=1.0))
+            return
+
+        results = tuneThresholdfromScore(scores, labels, [1, 0.1])
+        eer = results[1]
+        fnrs, fprs, thresholds = ComputeErrorRates(scores, labels)
+        p_trg, c_miss, c_fa = 0.05, 1, 1
```
Comment (Emrys365) [inline espnet2/train/spk_trainer.py:158]: Can you add some comments to explain the numbers (`[1, 0.1]` and `0.05, 1, 1`) here?

File: espnet2/train/spk_trainer.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        mindcf, _ = ComputeMinDcf(fnrs, fprs, thresholds, p_trg, c_miss, c_fa)
+        print("eer", eer, "mindcf", mindcf)
```
Comment (Emrys365) [inline espnet2/train/spk_trainer.py:160]:
```suggestion
```

File: espnet2/train/spk_trainer.py (hunk 14 above)
Comment (Jungjee) [reply espnet2/train/spk_trainer.py:158]: I will add in part 4. Hope you don't mind!

File: espnet2/train/spk_trainer.py (hunk 15 above)
Comment (Jungjee) [reply espnet2/train/spk_trainer.py:160]: This one too.

File: espnet2/train/spk_trainer.py (hunk 14 above)
Comment (Emrys365) [reply espnet2/train/spk_trainer.py:158]: OK. Just to comment here for convenience.

Comment (Jungjee) [reply espnet2/train/spk_trainer.py:158]: Done.

File: espnet2/train/preprocessor.py
Comment (sw005320) [inline espnet2/train/preprocessor.py:1]: This is more for a question.
Is there a way to keep the same random seed even in the multiple worker environment, so that we could reproduce the same augmented data?

Comment (sw005320) [reply espnet2/train/preprocessor.py:1]: Also, this extension is a very good example and I really appreciate both @Jungjee and @Emrys365’s efforts!

**Review by sw005320 — APPROVED** [review]
LGTM!

File: espnet2/train/preprocessor.py
Comment (Emrys365) [reply espnet2/train/preprocessor.py:1]: > This is more for a question. Is there a way to keep the same random seed even in the multiple worker environment, so that we could reproduce the same augmented data?

I think it is handled for the sequence iterator by the `worker_init_fn`:
https://github.com/espnet/espnet/blob/master/espnet2/iterators/sequence_iter_factory.py#L13-L17

Comment (sw005320) [reply espnet2/train/preprocessor.py:1]: I see, thx!
